keps/2349-multikueue-external-custom-job-support/README.md

```diff
@@ -0,0 +1,172 @@
+# KEP-2349: MultiKueue external custom Job support
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories ( Optional )](#user-stories--optional-)
+  - [Constraints](#constraints)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Risks](#risks)
+    - [Mitigations](#mitigations)
+- [Design Details](#design-details)
+  - [Generic Adapter via ConfigMap](#generic-adapter-via-configmap)
+  - [Workload-Only Sync for External Controllers](#workload-only-sync-for-external-controllers)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+ This KEP introduces a mechanism to support MultiKueue for custom job frome external controllers without requiring modifications changes in external job controllers.  This introduces a generic, built-in MultiKueue adapter that is configurable via a configmap.
+
+## Motivation
+
+Many Kueue adopters have specific CustomResources which might not be public.
+Kueue should give a possibility to manage such CustomResources across multiple clusters.
+
+### Goals
+
+Give a possibility for external controllers to manage private CustomResources in a MultiKueue environment.
+
+
+### Non-Goals
+
+## Proposal
+
+MultiKueue controller will be configurable via configmap. Based on this configmap, a generic MultiKueueAdapter will be implemented within Kueue. When processing a workload without supported built-in adapter, generic MultiKueueAdapter will be called based on GroupVersionKind (GVK) defined in configmap. It will use configuration defined to manage lifycycle's on the worker clusters.
+
+### User Stories ( Optional )
+Ability to schedule resources not yet supported by kueue like Tekton.
+
+### Constraints
+
+- Limited Transformation: Generic Adaptor are limited to simple JsonPatch transformation configurable.
+
+### Risks and Mitigations
+
+#### Risks
+
+ Configmaps are unstructured by default. They don't provide validation by default unlike CRD. User could make typo mistakes or give invalid syntax in configmap. Or user could give invalid syntax.
+
+#### Mitigations
+
+Configmap which aren't correct should logged like Invalid GVK format, Invalid JsonPatch ops. We could have cli command to validate adapter logic.
+
+## Design Details
+
+### Generic Adapter via ConfigMap
```

Comment:
Please also add a note that the mechanism would allow to replace implementation of the multiple built-in adapters, by registering the in-memory Generic Adapter for their GVKs.

For example I see no blocker to use it for JobSet, RayCluster, RayJob - it would be a great win to drop so much of code.

It would be a great side-effect win.

Reply:
> @mimowo out of curiosity, which adapters do you think it won't able to replace and why? 

Thank you for the question. The ones which are not fitting the generalized adapter:
1. Job
2. Pod

Reasons:
1. short term only, because we are using MultiKueueBatchJobWithManagedBy, this env variable will probably go away in a year or so when the feature in k8s reaches GA
2. because we re using schedulingGates instead of managedBy, unless we can generalize the solution to add / remove schedulingGates, but I think we can leave it out of scope for now

>  @khrm I think it would be useful to document what are the use cases that we know about that won't fit the proposed generic solution, so maintainers of external framework will have more info when they need to choose an implementation.

+1

Reply:
I have added this to KEP.